rclcpp/src/rclcpp/executors/detail/timer_manager.hpp

```diff
@@ -0,0 +1,356 @@
+          return timer_data == e.get();
+        }
+      );
+
+      if (it != all_timers.end()) {
+        all_timers.erase(it);
+      }
+      return true;
+    }
+    return false;
+  }
+
+  /**
+   * @brief adds the given timer_data to the map of running timers, if valid.
+   *
+   * Advances the rcl timer.
+   * Computes the next call time of the timer.
+   * readds the timer to the map of running timers
+   */
+  void add_timer_to_running_map(const TimerData * timer_data)
+  {
+    rcl_ret_t ret = rcl_timer_call(const_cast<rcl_timer_t *>(timer_data->rcl_ref.get()));
+    if (ret == RCL_RET_TIMER_CANCELED) {
+      return;
+    }
+
+    int64_t next_call_time;
+
+    ret = rcl_timer_get_next_call_time(timer_data->rcl_ref.get(), &next_call_time);
+
+    if (ret == RCL_RET_OK) {
+      running_timers.emplace(next_call_time, timer_data);
+    }
+
+    // wake up the timer thread so that it can pick up the timer
+    used_clock_for_timers.cancel_sleep_or_wait();
+  }
+
+  /**
+   * Returns the time when the next timer becomes ready
+   */
+  std::chrono::nanoseconds get_next_timer_ready_time() const
+  {
+    if (running_timers.empty()) {
+      // can't use std::chrono::nanoseconds::max, as wait_for
+      // internally computes end time by using ::now() + timeout
+      // as a workaround, we use some absurd high timeout
+      return std::chrono::nanoseconds(used_clock_for_timers.now().nanoseconds()) + std::chrono::hours(10000);
+    }
+    return running_timers.begin()->first;
+  }
+
+  void call_ready_timer_callbacks()
+  {
+    auto readd_timer_to_running_map = [this](TimerMap::node_type && e)
+      {
+        const auto & timer_data = e.mapped();
+        if(remove_if_dropped(timer_data))
+        {
+          return;
+        }
+
+        int64_t next_call_time;
+
+        auto ret = rcl_timer_get_next_call_time(timer_data->rcl_ref.get(), &next_call_time);
+
+        if (ret == RCL_RET_OK) {
+          e.key() = std::chrono::nanoseconds(next_call_time);
+          running_timers.insert(std::move(e));
+        }
+      };
+
+    while (!running_timers.empty()) {
+
+      if(remove_if_dropped(running_timers.begin()->second))
+      {
+        continue;
+      }
+
+      int64_t time_until_call;
+
+      const rcl_timer_t * rcl_timer_ref = running_timers.begin()->second->rcl_ref.get();
+      auto ret = rcl_timer_get_time_until_next_call(rcl_timer_ref, &time_until_call);
+      if (ret == RCL_RET_TIMER_CANCELED) {
+        running_timers.erase(running_timers.begin());
+        continue;
+      }
+
+      if (time_until_call <= 0) {
+        // advance next call time;
+        rcl_ret_t ret = rcl_timer_call(const_cast<rcl_timer_t *>(rcl_timer_ref));
+        if (ret == RCL_RET_TIMER_CANCELED) {
+          running_timers.erase(running_timers.begin());
+          continue;
+        }
+
+        // timer is ready, execute callback
+        running_timers.begin()->second->timer_ready_callback();
```

Comment:
a couple of comments on this:
 1. we should check somewhere that this is callable (e.g. when adding a timer)
 2. the current timers manager supports two modes of operation: in one mode, when a timer is ready it will trigger an event and the timer will then be executed by the events-executor thread; in the other mode the timers manager will directly execute the timer's user callback (without pushing any event). This second mode can give much more precision in the timer's execution and I see value in preserving it.
 
Maybe, to support the second mode, we should just set a `timer_ready_callback` that executes the timer...